state/machine.go

```diff
@@ -2230,6 +2273,24 @@ func removeUpgradeSeriesLockTxnOps(machineDocId string) []txn.Op {
 	}
 }
 
+func setUpgradeSeriesTxnOps(machineDocId string, unitIndex int, status params.UnitSeriesUpgradeStatus) []txn.Op {
+	unitField := fmt.Sprintf("prepareunits.%d.status", unitIndex)
+	return []txn.Op{
+		{
+			C:      machinesC,
+			Id:     machineDocId,
+			Assert: isAliveDoc,
+		},
+		{
+			C:      machineUpgradeSeriesLocksC,
+			Id:     machineDocId,
+			Assert: txn.DocExists,
+			Update: bson.D{
+				{"$set", bson.D{{unitField, status}}}},
```

Comment:
It would probably be better to have the unit fields keyed off the unit tag rather than an offset. This would make it easier to make the sested set. Also you should assert that the field matches the expected value.